### PR TITLE
[Numpy] add op nan_to_num

### PR DESCRIPTION
**Pull request by Alicia1529**
## Description ##
add numpy operator nan_to_num

### CODE REVIEW

File: python/mxnet/ndarray/numpy/_op.py [hunk 1]
```diff
@@ -5168,3 +5169,102 @@ def resize(a, new_shape):
...
+    >>> y
+    array([[-1.79769313e+308,  1.11111000e+005,  2.22222000e+005],
+       [ 2.22222000e+005,  2.22222000e+005, -1.79769313e+308]], dtype=float64)
+    """
+    if isinstance(x, numeric_types):
+        return _np.nan_to_num(x, copy, nan, posinf, neginf)
+    elif isinstance(x, NDArray):
+        if x.dtype in ['int8', 'uint8', 'int32', 'int64']:
+            return x
+        if copy is not True:
```
Comment (hgt312) [inline python/mxnet/ndarray/numpy/_op.py]:
```suggestion
        if not copy:
```

File: python/mxnet/symbol/numpy/_symbol.py [hunk 2]
```diff
@@ -4824,4 +4824,102 @@ def resize(a, new_shape):
...
+    out : ndarray
+        `x`, with the non-finite values replaced. If `copy` is False, this may
+        be `x` itself.
+
+    Notes
+    -----
+    NumPy uses the IEEE Standard for Binary Floating-Point for Arithmetic
+    (IEEE 754). This means that Not a Number is not equivalent to infinity.
+
+    Examples
```
Comment (hgt312) [inline python/mxnet/symbol/numpy/_symbol.py]: 'Examples' are not needed in `_symbol.py`.

File: src/operator/tensor/elemwise_unary_op.h [hunk 3]
```diff
@@ -660,6 +661,141 @@ void AroundOpForward(const nnvm::NodeAttrs& attrs,
...
+}
+
+template<int req>
+struct nan_to_num_backward {
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i,
+                                  DType* in_grad,
+                                  const DType* out_grad,
+                                  const DType* in_data) {
+    int val = 1*out_grad[i];
```
Comment (hgt312) [inline src/operator/tensor/elemwise_unary_op.h]:
```suggestion
    DType val = out_grad[i];
```

File: src/operator/tensor/elemwise_unary_op.h [hunk 4]
```diff
@@ -660,6 +661,141 @@ void AroundOpForward(const nnvm::NodeAttrs& attrs,
...
+
+template<typename xpu>
+void NumpyNanToNumOpBackward(const nnvm::NodeAttrs& attrs,
+                             const OpContext& ctx,
+                             const std::vector<TBlob>& inputs,
+                             const std::vector<OpReqType>& req,
+                             const std::vector<TBlob>& outputs) {
+  CHECK_EQ(inputs.size(), 2U);
+  CHECK_EQ(outputs.size(), 1U);
+  CHECK_NE(req[0], kWriteInplace);
```
Comment (hgt312) [inline src/operator/tensor/elemwise_unary_op.h]: Check dtype, if int types, raise an error.

File: src/operator/tensor/elemwise_unary_op.h [hunk 5]
```diff
@@ -660,6 +661,131 @@ void AroundOpForward(const nnvm::NodeAttrs& attrs,
...
+  CHECK_EQ(inputs.size(), 1U);
+  CHECK_EQ(outputs.size(), 1U);
+  CHECK_EQ(req.size(), 1U);
+  mshadow::Stream<xpu> *s = ctx.get_stream<xpu>();
+  const TBlob& in_data = inputs[0];
+  const TBlob& out_data = outputs[0];
+  const NumpyNanToNumParam& param = nnvm::get<NumpyNanToNumParam>(attrs.parsed);
+  using namespace mxnet_op;
+
+  if (!common::is_float(in_data.type_flag_)&&req[0] == kWriteInplace) return;
```
Comment (haojin2) [inline src/operator/tensor/elemwise_unary_op.h]:
```c++
if (!common::is_float(in_data.type_flag_) && req[0] == kWriteInplace) return;
```

File: src/operator/tensor/elemwise_unary_op.h [hunk 6]
```diff
@@ -660,6 +661,131 @@ void AroundOpForward(const nnvm::NodeAttrs& attrs,
...
+  if (!common::is_float(in_data.type_flag_)) return copy(s, out_data, in_data);
```
Comment (haojin2) [inline src/operator/tensor/elemwise_unary_op.h]:
```c++
if (!common::is_float(in_data.type_flag_)) {
  copy(s, out_data, in_data);
  return;
}
```

File: src/operator/tensor/elemwise_unary_op.h [hunk 7]
```diff
@@ -660,6 +661,131 @@ void AroundOpForward(const nnvm::NodeAttrs& attrs,
...
+    if (param.posinf.has_value()) {
+      posinf = static_cast<DType>(param.posinf.value());
+    } else {
+      posinf = mshadow::red::limits::MaxValue<DType>();
+    }
+    if (param.neginf.has_value()) {
+      neginf = static_cast<DType>(param.neginf.value());
+    } else {
+      neginf = mshadow::red::limits::MinValue<DType>();
+      }
```
Comment (haojin2) [inline src/operator/tensor/elemwise_unary_op.h]:
```c++
    if (param.neginf.has_value()) {
      neginf = static_cast<DType>(param.neginf.value());
    } else {
      neginf = mshadow::red::limits::MinValue<DType>();
    }
```

File: tests/python/unittest/test_numpy_op.py [hunk 8]
```diff
@@ -4164,6 +4164,108 @@ def hybrid_forward(self, F, x, *args, **kwargs):
...
 
 
+@with_seed()
+@use_np
+def test_np_nan_to_num():
+    def take_ele_grad(ele):
+        if _np.isinf(ele) or _np.isnan(ele):
+            return 0
+        return 1
+    def np_nan_to_num_grad(data):
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py]: one more blank line above this function.

File: tests/python/unittest/test_numpy_op.py [hunk 9]
```diff
@@ -4164,6 +4164,108 @@ def hybrid_forward(self, F, x, *args, **kwargs):
...
+            mx_out_gluon.backward()
+            assert_almost_equal(x2.grad.asnumpy(), expected_grad, rtol=1e-3, atol=1e-5)
+                                        
+        # Test imperative once again
+        # if copy = False, the value of x1 and x2 has changed
+        if copy == True:            
+            np_out = _np.nan_to_num(x1)
+            mx_out = np.nan_to_num(x3)
+            assert_almost_equal(mx_out.asnumpy(), np_out, rtol=1e-3, atol=1e-5, use_broadcast=False)
+
```
Comment (haojin2) [inline tests/python/unittest/test_numpy_op.py]: only 2 blank lines between tests here.

**Pull request merged**